test/common/ssl/ssl_socket_test.cc

```diff
@@ -678,6 +678,159 @@ TEST_P(SslSocketTest, FailedClientAuthSanVerification) {
            GetParam());
 }
 
+TEST_P(SslSocketTest, FailedClientCertificateDefaultExpirationVerification) {
+  std::string client_ctx_json = R"EOF(
+  {
+    "cert_chain_file": "{{ test_rundir }}/test/common/ssl/test_data/expired_san_uri_cert.pem",
+    "private_key_file": "{{ test_rundir }}/test/common/ssl/test_data/expired_san_uri_key.pem"
+  }
+  )EOF";
+
+  std::string server_ctx_json = R"EOF(
+  {
+    "cert_chain_file": "{{ test_tmpdir }}/unittestcert.pem",
+    "private_key_file": "{{ test_tmpdir }}/unittestkey.pem",
+    "ca_cert_file": "{{ test_rundir }}/test/common/ssl/test_data/ca_cert.pem"
+  }
+  )EOF";
+
+  testUtil(client_ctx_json, server_ctx_json, "", "spiffe://lyft.com/test-team", "", "", "", "",
+           "ssl.fail_verify_error", false, GetParam());
+}
+
+TEST_P(SslSocketTest, FailedClientCertificateExpirationVerification) {
+  envoy::api::v2::Listener listener;
+  envoy::api::v2::listener::FilterChain* filter_chain = listener.add_filter_chains();
+  envoy::api::v2::auth::TlsCertificate* server_cert =
+      filter_chain->mutable_tls_context()->mutable_common_tls_context()->add_tls_certificates();
+  server_cert->mutable_certificate_chain()->set_filename(
+      TestEnvironment::substitute("{{ test_tmpdir }}/unittestcert.pem"));
+  server_cert->mutable_private_key()->set_filename(
+      TestEnvironment::substitute("{{ test_tmpdir }}/unittestkey.pem"));
+  envoy::api::v2::auth::CertificateValidationContext* server_validation_ctx =
+      filter_chain->mutable_tls_context()
+          ->mutable_common_tls_context()
+          ->mutable_validation_context();
+  server_validation_ctx->mutable_trusted_ca()->set_filename(
+      TestEnvironment::substitute("{{ test_rundir }}/test/common/ssl/test_data/ca_cert.pem"));
+
+  // Enforce expiration checks (cheked by default as well).
+  server_validation_ctx->set_allow_expired_certificate(false);
+
+  envoy::api::v2::auth::UpstreamTlsContext client;
+  envoy::api::v2::auth::TlsCertificate* client_cert =
+      client.mutable_common_tls_context()->add_tls_certificates();
+  client_cert->mutable_certificate_chain()->set_filename(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/common/ssl/test_data/expired_san_uri_cert.pem"));
+  client_cert->mutable_private_key()->set_filename(TestEnvironment::substitute(
+      "{{ test_rundir }}/test/common/ssl/test_data/expired_san_uri_key.pem"));
+
+  testUtilV2(listener, client, "", false, "", "", "spiffe://lyft.com/test-team", "",
+             "ssl.fail_verify_error", 1, GetParam());
+}
+
+TEST_P(SslSocketTest, ClientCertificateExpirationAllowedVerification) {
```

Comment:
I didn't look too carefully, but it looks like this test is identical to the previous one, except the value for set_allow_expired_certificate and the expected result. Can the two share a single function taking a boolean? If it's messy, don't worry about it.

Reply:
They could share a single function taking a boolean if it was part of `SslSocketTest` because it needs access to `GetParam()`.  Seems a bit messy when only two methods need this.

However, I could have a utility function so it's reduced to:

```
TEST_P(SslSocketTest, FailedClientCertificateExpirationVerification) {
  envoy::api::v2::Listener listener;
  envoy::api::v2::auth::UpstreamTlsContext client;

  configureExpiredClientCertificate(listener, client, false);

  testUtilV2(listener, client, "", false, "", "", "spiffe://lyft.com/test-team", "",
             "ssl.fail_verify_error", 1, GetParam());
}
```